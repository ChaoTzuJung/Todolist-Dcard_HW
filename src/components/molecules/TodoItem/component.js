import React from 'react';
import classnames from 'classnames';
import List from 'components/molecules/List';
import TodoPanel from 'components/molecules/TodoPanel';

import styles from './index.css';

const TodoItem = ({
	id,
	text,
	deadline,
	file,
	comment,
	checked,
	star,
	edit,
	...other,
}) => (
	<div className={classnames(styles.todoItem, styles.todo)}>
		<List
			id={id}
			text={text}
			deadline={deadline}
			file={file}
			comment={comment}
			checked={checked}
			star={star}
			edit={edit}
			{...other}
		/>
		{edit && (<TodoPanel />)}
	</div>
);

export default TodoItem;
