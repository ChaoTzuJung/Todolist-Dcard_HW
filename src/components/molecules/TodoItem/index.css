.todo-item {
	position: relative;
	margin-top: 8px;

	@media screen and (--phone) {
		margin: auto;
		width: 75%;
		margin-top: 16px;
	}

	& > .delete-item {
		display: flex;
		align-items: center;

		& > span {
			display: flex;
			align-items: center;
			justify-content: center;
			position: absolute;
			right: -64px;
			width: 48px;
			height: 48px;
			border-radius: 50%;
			background: var(--lightGray);
			box-shadow: 0 4px 4px 0 #c8c8c8;
			cursor: pointer;

			@media screen and (--small-tablet) {
				right: -48px;
				width: 32px;
				height: 32px;
			}

			@media screen and (--phone) {
				width: 24px;
				height: 24px;
				border-radius: 0%;
				background: transparent;
				box-shadow: none;
				right: -34px;
			}

			& > i {
				color: var(--red);
				font-size: 22px;

				@media screen and (--small-tablet) {
					font-size: 18px;
				}

				@media screen and (--phone) {
					font-size: 20px;
				}
			}

			& > .deleted {
				color: var(--red);
			}
		}
	}
}
