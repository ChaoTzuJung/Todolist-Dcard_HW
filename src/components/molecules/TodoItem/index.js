import React, { Component } from 'react';
import classnames from 'classnames';
import { isExist } from 'util/helper';
import List from 'components/molecules/List';
import TodoPanel from 'components/molecules/TodoPanel';

import styles from './index.css';

class TodoItem extends Component {
	constructor(props) {
		super(props);
		this.input = React.createRef();
		this.panel = React.createRef();

		this.state = {
			todo: {
				text: '',
				checked: false,
				star: false,
				edit: true,
				startTime: '',
				date: '',
				time: '',
				fileData: '',
				fileName: '',
				fileType: '',
				hasImage: false,
				textarea: '',
			},
		};
		this.handleSave = this.handleSave.bind(this);
		this.handleCancel = this.handleCancel.bind(this);
		this.addStar = this.addStar.bind(this);
		this.onEdit = this.onEdit.bind(this);
		this.onCheck = this.onCheck.bind(this);
	}

	handleSave(data) {
		// data 是 在 panel Add task 時 所戴的資料
		console.log(data);
		const { date, time, startTime, fileData, fileName, fileType, hasImage, textarea } = data;
		//  List contains a simple text input. We need to expose a way for other components to read the text written in the List component.
		const todo = {};
		if (isExist(this.input.current.state.text)) {
			this.setState(prevState => ({
				todo: {
					...prevState.todo,
					text: this.input.current.state.text,
					edit: false,
					startTime,
					date,
					time,
					fileData,
					fileName,
					fileType,
					hasImage,
					textarea,
					deadline: isExist(startTime),
					file: isExist(fileData),
					comment: isExist(textarea),
				},
			}));
		} else {
			alert('請輸入 Todo 標題 !');
		}
	}

	handleCancel() {
		this.setState(prevState => ({
			todo: {
				...prevState.todo,
				edit: false,
			},
		}));
	}

	addStar() {
		this.setState(prevState => ({
			todo: {
				...prevState.todo,
				star: !prevState.todo.star,
			},
		}));
	}

	onEdit() {
		this.setState(prevState => ({
			todo: {
				...prevState.todo,
				edit: !prevState.todo.edit,
			},
		}));
	}

	onCheck(value, checked) {
		console.log(value);
		console.log(checked);
		this.setState(prevState => ({
			todo: {
				...prevState.todo,
				// checked: !prevState.todo.checked,
				checked,
			},
		}));
	}

	render() {
		const { todo } = this.state;
		const { className, edit } = this.props;
		const todoData = { ...todo };

		return (
			<div className={classnames(styles.todoItem, className)}>
				<List
					text={todo.text}
					checked={todo.checked}
					star={todo.star}
					edit={edit}
					deadline={todo.date || todo.time}
					date={`${todo.date} ${todo.time}` || todo.date || todo.time}
					file={todo.file}
					comment={todo.comment}
					ref={this.input}
					addStar={this.addStar}
					onEdit={this.onEdit}
					onChangeCheckbox={this.onCheck}
				/>
				{edit && (
					<TodoPanel
						// Add task and edit
						todoData={todoData}
						onSave={this.handleSave}
						onCancel={this.handleCancel}
						ref={this.panel}
					/>
				)}
			</div>
		);
	}
}

export default TodoItem;
