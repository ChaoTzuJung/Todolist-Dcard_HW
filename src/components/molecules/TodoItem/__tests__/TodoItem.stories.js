import React from 'react';
import { withKnobs, text, boolean } from '@storybook/addon-knobs/react';

import TodoItem from 'components/molecules/TodoItem/component';

import { storiesOf } from '@storybook/react';

const stories = storiesOf('Commons|molecules/TodoItem', module);

stories.addDecorator(withKnobs);

stories.add('__interactive', () => (
	<div style={{ margin: '100px' }}>
		<TodoItem
			id={Date.now()}
			text={text('text', 'Type Something Hereâ€¦')}
			date={text('deadline', '')}
			file={text('file', '')}
			comment={text('comment', '')}
			checked={boolean('checked', false)}
			star={boolean('star', false)}
			edit={boolean('edit', true)}
		/>
	</div>
));
