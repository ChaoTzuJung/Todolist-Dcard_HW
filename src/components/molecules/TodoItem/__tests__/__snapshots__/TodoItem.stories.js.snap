// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<div
  style={
    Object {
      "margin": "100px",
    }
  }
>
  <TodoItem>
    <div
      className="todoItem"
    >
      <Contain(List)
        addStar={[Function]}
        comment={false}
        completed={false}
        deadline={false}
        file={false}
        handleCheckboxChange={[Function]}
        onEdit={[Function]}
        star={false}
      >
        <List
          addStar={[Function]}
          cacheTodo={Object {}}
          comment={false}
          completed={false}
          deadline={false}
          file={false}
          handleChange={[Function]}
          handleCheckboxChange={[Function]}
          onEdit={[Function]}
          star={false}
        >
          <div
            className="list"
          >
            <div
              className="left"
            >
              <div
                className="top"
              >
                <Checkbox
                  checked={false}
                  className=""
                  onChange={[Function]}
                >
                  <label
                    className="label"
                  >
                    <input
                      checked={false}
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="checkbox"
                    >
                      <Icon>
                        <i
                          className="icon material-icons"
                        >
                          check
                        </i>
                      </Icon>
                    </span>
                  </label>
                </Checkbox>
              </div>
              <div
                className="down"
              />
            </div>
            <div
              className="right"
            >
              <span
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <Icon
                  className=""
                >
                  <i
                    className="icon material-icons"
                  >
                    star_border
                  </i>
                </Icon>
              </span>
              <span
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <Icon
                  className=""
                  outlined={true}
                >
                  <i
                    className="icon material-icons-outlined"
                  >
                    edit
                  </i>
                </Icon>
              </span>
            </div>
          </div>
        </List>
      </Contain(List)>
    </div>
  </TodoItem>
</div>
`;
